### ISSUE TITLE
Expose Waypoint client/api as a library

### ISSUE THREAD

**Issue by eveld** [hashicorp/waypoint#3118]
I am currently building an integration with Waypoint triggers and have to craft the requests myself. 

It would be really useful if there was a client/api package in the Waypoint codebase that I could import to communicate with the Waypoint server. This client could then also have prescriptive ways for authentication, etc.

**Comment by mitchellh** [comment]
We should probably generate a client lib from protos and build up additional functionality in `waypoint-plugin-sdk` and not in this repo since this repo is never meant to be imported as a library by 3rd parties.

**Comment by izaaklauer** [comment]
Good point from @catsby  - maybe the client should live in it's own repo. It might feel weird to import the "waypoint-plugin-sdk" to get a client unrelated to plugins.

**Comment by mitchellh** [comment]
I don't believe we should extract `internal/client`. I purpose made `internal/client` internal to avoid some of the challenges we've had in our other projects (Consul, Vault as a personal paint point) where the public API client was also the internal API client. 

The primary problem I discovered with that pattern is that the consumer is two separate personas: the core team doing core things and the general user doing product extension/integration. For example, once a client is extracted, it itself can no longer depend on internal packages (without also extracting those). 

As a concrete example of how this would affect Waypoint: our `internal/client` is also currently responsible for spinning up a local runner ("just-in-time") for local operation execution. The runner package is internal, we don't really want people building custom runners on top of that package (because it itself depends on a bunch of internal packages/functionality). For 99% of "general users", they want the client package to be a gRPC API client, _not_ a runner. So (1) the runner code would have to be extracted (2) the runner code would add a ton of "bloat" to the normal user use case (3) all transitive internal dependencies would have to be extracted.

Or on the flip side as a devil's advocate, you'd have to rearchitect the entire `internal/client` package so that its more flexible (i.e. interface-driven runner execution instead of embedding `internal/runner`). There's no practical reason for this today except to try to share that package. 

From my experience on other projects, I've come to the conclusion its just easier to maintain two API clients: a public one, and an internal one. The public one is polished, aimed at HTTP/gRPC API consumption, and focused on user oriented workflows. The internal one can be more raw and focused on _product implementation_.

So my proposal would be to generate an entirely new client. I'm supportive of @catsby's suggestion of making it a sub-package of the SDK.
